### ISSUE TITLE
libvirt__network_name doesn't work

### ISSUE THREAD

**Issue by nergdron** [vagrant-libvirt/vagrant-libvirt#1202]
### Steps to reproduce
1. define a default private network block using :libvirt_network_name to set a network name.

### Expected behaviour
network should be created in libvirt with the specified name.

### Actual behaviour
network is created with other custom settings, but with the generic name "vagrant-private-dhcp" or similar. this also causes collisions between multiple Vagrantfiles, since even if you define a custom network name, they try to stomp on each other and every run after the first errors out.

### System configuration
**OS/Distro version:**: Ubuntu 20.10

**Libvirt version:** 6.6.0-1ubuntu3.1

**Output of `vagrant version; vagrant plugin list`:**
```
Installed Version: 2.2.14
Latest Version: 2.2.14
 
You're running an up-to-date version of Vagrant!
vagrant-libvirt (0.3.0, global)
  - Version Constraint: > 0
```

**Attach Output of `VAGRANT_LOG=debug vagrant ... --provider=libvirt >vagrant.log 2>&1`**
```
DEBUG create_networks: In config found network type private_network options {:type=>"dhcp", :mode=>"nat", :libvirt__network_name=>"private-test", :libvirt__network_address=>"10.11.12.0", :protocol=>"tcp", :id=>"e9d889b6-0d38-4f77-ae40-8df7a0506244"}
DEBUG create_networks: In config found network type forwarded_port options {:guest=>22, :host=>2222, :host_ip=>"127.0.0.1", :id=>"ssh", :auto_correct=>true, :protocol=>"tcp"}
DEBUG create_networks: Searching for network with options {:iface_type=>:private_network, :network_name=>"vagrant-libvirt", :ip=>"192.168.121.0", :netmask=>"255.255.255.0", :dhcp_enabled=>true, :forward_mode=>"nat", :guest_ipv6=>"yes", :autostart=>false, :bus=>nil, :slot=>nil}
DEBUG create_networks: looking up network with ip == 192.168.121.0
DEBUG create_networks: Checking that network name does not clash with ip
DEBUG create_networks: looking up network named vagrant-libvirt
DEBUG create_networks: generating name for bridge
DEBUG create_networks: looking up bridge named virbr0
DEBUG create_networks: looking up bridge named virbr1
DEBUG create_networks: found available bridge name virbr1
DEBUG create_networks: created network
 INFO create_networks: Saving information about created network vagrant-libvirt, UUID=e9680da4-3b18-4ec2-b3f7-6f5c951c05ae to file /tmp/.vagrant/machines/test/libvirt/created_networks.
DEBUG create_networks: Searching for network with options {:iface_type=>:private_network, :netmask=>"255.255.255.0", :dhcp_enabled=>true, :forward_mode=>"nat", :type=>"dhcp", :mode=>"nat", :libvirt__network_name=>"private-test", :libvirt__network_address=>"10.11.12.0", :protocol=>"tcp", :id=>"e9d889b6-0d38-4f77-ae40-8df7a0506244", :network_name=>"vagrant-private-dhcp", :network_address=>"10.11.12.0"}
DEBUG create_networks: looking up network with ip == 10.11.12.0
DEBUG create_networks: generating name for bridge
DEBUG create_networks: looking up bridge named virbr0
DEBUG create_networks: looking up bridge named virbr1
DEBUG create_networks: looking up bridge named virbr2
DEBUG create_networks: found available bridge name virbr2
DEBUG create_networks: created network
 INFO create_networks: Saving information about created network vagrant-private-dhcp, UUID=bcb98d19-c3f4-426b-8e2a-9c52c7052f6a to file /tmp/.vagrant/machines/test/libvirt/created_networks.
```

**A Vagrantfile to reproduce the issue:**
```
Vagrant.configure("2") do |config|
  config.vm.box = "generic/ubuntu2004"

  config.vm.network :private_network,
    :type => "dhcp",
    :mode => "nat",
    :libvirt__network_name => "private-test",
    :libvirt__network_address => "10.11.12.0"

  config.vm.define "test" do |test|
    test.vm.hostname = "test"
  end
end
```

**Are you using upstream vagrant package or your distros package?**
Upstream PPA

**Additional notes:**

* other provider specific options like `:libvirt__network_address` seem to work fine.
* the management network option `management_network_name` also seems to work fine.
* doesn't seem to matter if the private network is defined globally in the Vagrantfile or local to the specific box, the network name never gets set.
* debug log shows that it isn't even passing the value from the variable into its libvirt calls, it's just using that default network name.
* the docs indicate the default network it connects to is called "default", but that's obviously incorrect since it always ends up being "vagrant-private-dhcp".

**Comment by electrofelix** [comment]
I think I've identified the issue. It appears there are 2 hardcoded references to `vagrant-private-dhcp`.
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/978e92eac5c4be8b662c98e62e60166f671d455b/lib/vagrant-libvirt/util/network_util.rb#L106-L108

and 

https://github.com/vagrant-libvirt/vagrant-libvirt/blob/978e92eac5c4be8b662c98e62e60166f671d455b/lib/vagrant-libvirt/action/create_networks.rb#L249-L259

I think the main issue is that it was missed with the util file that the main purpose of the scoped_hash_override is to covert `libvirt__<option>` to `<option>` in the returned hash:
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/978e92eac5c4be8b662c98e62e60166f671d455b/lib/vagrant-libvirt/util/network_util.rb#L91-L96

This should mean that `libvirt__network_name` should become `network_name`. So a simple check on whether `options[:network_name]` was nil before setting it to `vagrant-private-dhcp` would allow an override setting to be passed though, and then it just needs to be respected in the create_networks.rb code.

Cleaning out the remaining `libvirt__<option>` references in favour of `<option>` references instead would result in the following patch:
```
diff --git a/lib/vagrant-libvirt/action/create_networks.rb b/lib/vagrant-libvirt/action/create_networks.rb
index 9ca21d3..365dc3e 100644
--- a/lib/vagrant-libvirt/action/create_networks.rb
+++ b/lib/vagrant-libvirt/action/create_networks.rb
@@ -139,7 +139,8 @@ module VagrantPlugins
           @interface_network[:network_address] = net_address
 
           # if network is veryisolated, search by name
-          network = if @options[:libvirt__forward_mode] == 'veryisolated'
+          # may be set via :libvirt__forward_mode
+          network = if @options[:forward_mode] == 'veryisolated'
                       lookup_network_by_name(@options[:network_name])
                     elsif net_address
                       # otherwise, search by ip (if set)
@@ -213,12 +214,12 @@ module VagrantPlugins
         def handle_network_name_option(env)
           return if @options[:ip] || \
                     !@options[:network_name] || \
-                    !@options[:libvirt__forward_mode] == 'veryisolated'
+                    !@options[:forward_mode] == 'veryisolated'
 
           network = lookup_network_by_name(@options[:network_name])
           @interface_network = network if network
 
-          if @options[:libvirt__forward_mode] == 'veryisolated'
+          if @options[:forward_mode] == 'veryisolated'
             # if this interface has a network address, something's wrong.
             if @interface_network[:network_address]
               raise Errors::NetworkNotAvailableError,
@@ -247,7 +248,7 @@ module VagrantPlugins
         end
 
         def handle_dhcp_private_network(env)
-          net_address = @options[:libvirt__network_address]
+          net_address = @options[:network_address]
           net_address = '172.28.128.0' unless net_address
           network = lookup_network_by_ip(net_address)
 
@@ -255,7 +256,7 @@ module VagrantPlugins
 
           # Do we need to create new network?
           unless @interface_network[:created]
-            @interface_network[:name] = 'vagrant-private-dhcp'
+            @interface_network[:name] = @options[:network_name]
             @interface_network[:network_address] = net_address
 
             # Set IP address of network (actually bridge). It will be used as
diff --git a/lib/vagrant-libvirt/util/network_util.rb b/lib/vagrant-libvirt/util/network_util.rb
index ea71cb1..b5b4f30 100644
--- a/lib/vagrant-libvirt/util/network_util.rb
+++ b/lib/vagrant-libvirt/util/network_util.rb
@@ -103,7 +103,7 @@ module VagrantPlugins
               forward_mode: 'nat'
             }.merge(options)
 
-            if options[:type].to_s == 'dhcp' && options[:ip].nil?
+            if options[:type].to_s == 'dhcp' && options[:ip].nil? && options[:network_name].nil?
               options[:network_name] = 'vagrant-private-dhcp'
             end
```

I'm not entirely sure why it's currently written to use `vagrant-private-dhcp` instead of `default`, so I'd be a little hesitant to fix that. 

Possibly the correct behaviour should be based on the following list, taking the first option that matches:
* if network name is not nil => use provided
* if network address is provided => generate a name 'vagrant-private-dhcp' (if the address matches another network, that will be used instead when actually configurating)
* if dhcp is enabled and ip address is nil => use `default`

In other words it should fallback to default if no network address is defined, as otherwise it will need some name it can use to create the network definition under.


I'm a little hesitant to apply the above patch and release, purely because there are absolutely no tests on the existing network code, neither the util part nor the create networks action.

So I think the first step is to start with those and that hopefully will help at least confirm the existing behaviour so that after making the changes it's not breaking what is working for others unexpectedly.

**Comment by electrofelix** [comment]
I believe this has been fixed with #1248, so it should be resolved with the next release
